lib/chainlock/chainlock.js

```diff
@@ -126,28 +126,68 @@ ChainLock.fromObject = function fromObject(obj) {
   const data = ChainLock._fromObject(obj);
   return new ChainLock(data);
 };
+
 /**
  * Verify that the signature is valid against the Quorum using quorumPublicKey
  * @param {QuorumEntry} quorumEntry - quorum entry to test signature against
- * @returns {Promise<Boolean>} - if the signature is valid for this quorum
+ * @returns {Promise<Boolean>} - returns the result of the signature verification
  */
 async function verifySignatureAgainstQuorum(quorumEntry) {
   const { signature } = this;
   const { quorumPublicKey } = quorumEntry;
-  const signatureId = this.getSignatureIDForQuorumEntry(quorumEntry);
+  const signHash = this.getSignHashForQuorumEntry(quorumEntry);
 
   const blsInstance = await bls.getInstance();
 
   const quorumPubKey = blsInstance.PublicKey.fromBytes(Buffer.from(quorumPublicKey, 'hex'));
 
-  const aggregationInfo = blsInstance.AggregationInfo.fromMsgHash(quorumPubKey, signatureId);
+  const aggregationInfo = blsInstance.AggregationInfo.fromMsgHash(quorumPubKey, signHash);
   const thresholdSignature = blsInstance.Signature.fromBytes(Buffer.from(signature, 'hex'));
 
   thresholdSignature.setAggregationInfo(aggregationInfo);
 
   return thresholdSignature.verify();
 }
+
 ChainLock.prototype.verifySignatureAgainstQuorum = verifySignatureAgainstQuorum;
+
+/**
+ * Tries verifying first with the default offset, then with 0 offset, then with double offset
+ * @param {SimplifiedMNListStore} SMLStore - used to reconstruct quorum lists
+ * @param {number} offset - we use this offset as a starting height to identify the signatory
+ * which are not 100% at the chain tip.
+ * @returns {Promise<Boolean>}
+ */
+async function verifySignatureWithQuorumOffset(SMLStore, offset) {
+  const requestId = this.getRequestId();
+  const candidateSignatoryQuorum = this.selectSignatoryQuorum(SMLStore, requestId, offset);
+
+  let result = await this.verifySignatureAgainstQuorum(candidateSignatoryQuorum);
+
+  if (!result && offset === constants.LLMQ_SIGN_HEIGHT_OFFSET) {
```

Comment:
Could we move the comments that describe logic from JSDoc to exact code lines?

Reply:
done!